text/0195-associated-items.md

```diff
@@ -1190,6 +1197,73 @@ trait Mappable
 While the above demonstrates the versatility of associated types and `where`
 clauses, it is probably too much of a hack to be viable for use in `libstd`.
 
+### Associated consts in generic code
+
+There are some restrictions on uses of associated consts in generic code. These
+might be loosened or removed in the future (see the related sub-sections in
+"Unresolved questions" below).
+
+ 1. Values of constant expressions in match patterns cannot depend on a type
+    parameter (by extension, neither can the types of such expressions). This
+    restriction is necessary for exhaustiveness and reachability to be checked
+    in generic code.
+
+    Note that the dependence of a value on a type parameter may be indirect:
+
+    ```rust
+    enum MyEnum {
+        Var1,
+        Var2,
+    }
+    trait HasVar {
+        const VAR: MyEnum;
+    }
+    fn do_something<T: HasVar>(x: MyEnum) {
+        const y: MyEnum = <T>::VAR;
```

Comment:
Ah, right, I forgot to address that. My standpoint is basically that there's not much reason to apply that restriction to constants. It doesn't seem to be "official" in the sense of being mentioned in an RFC (?). It also wasn't applied to statics until rust-lang/rust#9186, and the motivation behind that change is not really applicable to constants, which are inlined.

It looks to me like that constants were forbidden from being type parameters as a side effect of the restriction on statics, and before now having generic constants hasn't actually been useful enough to matter anyway. This example doesn't require a const to be declared this way, but the latter discussion on array sizes really falls apart without that ability. You basically have to depend on the `TypeAdd` strategy if you want to do any kind of arithmetic to produce an array size. (Which maybe is OK; dealing with array sizes generically is already kind of icky in ways that associated consts alone won't fix, so it's just more motivation to figure that out.)